tests/cbmc/proofs/s2n_stuffer_write/s2n_stuffer_write_harness.c

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/s2n.h"
+#include "stuffer/s2n_stuffer.h"
+
+#include <assert.h>
+#include <sys/param.h>
+#include <cbmc_proof/cbmc_utils.h>
+#include <cbmc_proof/make_common_datastructures.h>
+#include <cbmc_proof/proof_allocators.h>
+
+void s2n_stuffer_write_harness() {
+    /* Non-deterministic inputs. */
+    struct s2n_stuffer *stuffer = cbmc_allocate_s2n_stuffer();
+    __CPROVER_assume(s2n_stuffer_is_valid(stuffer));
+    struct s2n_blob *blob = cbmc_allocate_s2n_blob();
+    __CPROVER_assume(s2n_blob_is_valid(blob));
+    uint32_t index;
+
+    /* Non-deterministically set initialized (in s2n_mem) to true. */
+    if(nondet_bool()) {
+        s2n_mem_init();
+    }
+
+    /* Save previous state from stuffer. */
+    struct s2n_stuffer old_stuffer = *stuffer;
+
+/**
+ * There should not be any constraints over blob.size, but that leads to an spurious
+ * unsigned overflow in the following assumption; therefore, we ignore this
+ * check here to keep blob.size as non-deterministic as possible.
+ */
+#pragma CPROVER check push
+#pragma CPROVER check disable "unsigned-overflow"
+    __CPROVER_assume(index < stuffer->blob.size && (index < old_stuffer.write_cursor ||
```

Comment:
I'm a bit skeptical when we have to disable warnings to make things work. If I understand correctly, assuming a non-zero sized blob then index should always be less than the size. And the fact that `size` is `uint32_t` then `index`'s max value will be `UINT32_MAX - 1`.

In the case where `size` is `0`, reading `stuffer->blob.data[index]` will result in garbage data no matter what the value of `index`.

Reply:
the overflow here is not in the index, but in the sum `old_stuffer.write_cursor + size`, since we assume both values as nondet numbers; I'll re-write and remove the pragma.

Reply:
We can use an intrinsic function from CBMC to check for overflow and then avoid it.